tools/make_patch.c

```diff
@@ -0,0 +1,944 @@
+#include <stdlib.h>
+#include <stdio.h>
+#include <stdint.h>
+#include <errno.h>
+#include <string.h>
+#include <ctype.h>
+
+struct symbol {
+	struct symbol *next;
+	unsigned int value;
+	char name[];
+};
+
+struct patch {
+	unsigned int offset;
+	unsigned int size;
+};
+
+struct buffer {
+	size_t size;
+	char buffer[];
+};
+
+struct asmfile {
+	struct asmfile *next;
+	struct symbol *symbols;
+	char name[];
+};
+
+struct asmfile *asmfiles = NULL;
+
+void free_buffer(void *buf)
+{
+	struct buffer *buffer = buf - sizeof(struct buffer);
+	free(buffer);
+}
+
+void *create_buffer()
+{
+	struct buffer *buffer = malloc(sizeof(struct buffer) + 0x100);
+	if (!buffer) {
+		fprintf(stderr, "Error: Cannot allocate memory\n");
+		return NULL;
+	}
+
+	buffer->size = 0x100;
+	return buffer->buffer;
+}
+
+void *expand_buffer(void *buf, const size_t size)
+{
+	struct buffer *buffer = buf - sizeof(struct buffer);
+
+	if (size > buffer->size) {
+		buffer = realloc(buffer, (buffer->size += 0x100));
+		if (!buffer) {
+			free_buffer(buf);
+			fprintf(stderr, "Error: Cannot allocate memory\n");
+			return NULL;
+		}
+	}
+
+	return buffer->buffer;
+}
+
+void free_list(void *list)
+{
+	// Frees an entire linked list
+	// Expects the ->next entry to be the first element!
+
+	while (list) {
+		void *next = *(void **)list;
+		free(list);
+		list = next;
+	}
+}
```

Comment:
```suggestion
void free_symbols(struct symbol *list)
{
	while (list) {
		struct symbol *next = list->next;
		free(list);
		list = next;
	}
}
```

Reply:
Resolved in https://github.com/pret/pokecrystal/pull/882/commits/9ff133bc78a23561f61b2597e36456f307771a29